docs/proposals/20210722-yurtcluster-operator.md

```diff
@@ -72,43 +74,174 @@ This is an operator, with CRD and Controller.
 CRD's definition:
 
 ```go
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+// ImageMeta allows to customize the image used for components that are not
+// originated from the openyurtio/openyurt release process
+type ImageMeta struct {
+	// ImageRepository sets the container registry to pull images from.
+	// if not set, the ImageRepository defined in YurtClusterSpec will be used instead.
+	// +optional
+	ImageRepository string `json:"imageRepository,omitempty"`
+	// ImageTag allows to specify a tag for the image.
+	// if not set, the tag releated to the YurtVersion defined in YurtClusterSpec will be used instead.
+	// +optional
+	ImageTag string `json:"imageTag,omitempty"`
+}
+
+// ComponentConfig defines the common config for the yurt components
 type ComponentConfig struct {
-        Enabled bool   `json:"enebled,omitempty"`
-        Version string `json:"version,omitempty"`
+	// Enabled indicates whether the yurt component has been enabled
+	// +optional
+	Enabled *bool `json:"enabled,omitempty"`
+	// ImageMeta allows to customize the image used for the yurt component
+	// +optional
+	ImageMeta `json:",inline"`
+}
+
+// YurtHub defines the configuration for yurthub component
+type YurtHub struct {
+	// ComponentConfig defines the common config for the yurt components
+	// +optional
+	ComponentConfig `json:",inline"`
+	// EnableResourceFilter enables to filter response that comes back from reverse proxy
+	// +optional
+	EnableResourceFilter *bool `json:"enableResourceFilter,omitempty"`
+	// AccessServerThroughHub enables pods access kube-apiserver through yurthub or not
+	// +optional
+	AccessServerThroughHub *bool `json:"accessServerThroughHub,omitempty"`
+	// AutoRestartNodePod represents whether to automatically restart the pod on the node after it was converted
+	// This will trigger the re-injection of environment variables (e.g. KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT)
+	// This should only be set when EnableResourceFilter is true.
+	// +optional
+	AutoRestartNodePod *bool `json:"autoRestartNodePod,omitempty"`
```

Comment:
How about add another two fields `PodManifestPath` and `KubeadmConfPath` in order to convert or revert node?

Reply:
Emm, these two fields seem to can be read from flags of the `kubelet` service. 
We can automatically extract it by regular expressions.

Reply:
`KubeadmConfPath` is used to find service config file of kubelet, so we can not extract it through kubelet service.

Reply:
Done!